Hair Rererereredux

Vimas21 opened the pull request:
🆑 SammasC123
rscadd: Adds capilliumate and hair dye as reagents. Capilliumate grows hair, and hair dye changes a persons hair color (facial and head). Works on contact.
/🆑

Vimas21 committed:
Hair Rerereredux

I wasn't kidding, I will
Added Side-Effect and Chaos

Hubblenaut commented:
Noting that crayons cannot be grinded right now. Not sure if there is any other way to acquire crayon dust.

SierraKomodo commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,100 @@
+		var/mob/living/carbon/human/H = M
+		if(prob(10))
+			to_chat(M, SPAN_NOTICE("Your tongue feels... fuzzy."))
+		if(H.stat != UNCONSCIOUS)
```
I'm not seeing the relation between the mob being unconscious and applying slurring - Why wouldn't the effect apply even if they're unconscious?

SierraKomodo commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,100 @@
+		var/datum/sprite_accessory/facial_hair/newbeard = /datum/sprite_accessory/facial_hair/vlongbeard
+		M.change_hair(initial(newhair.name))
+		M.change_facial_hair(initial(newbeard.name))
+		to_chat(M, SPAN_NOTICE("Your hair grows to extraordinary lengths!"))
```
Use `visible_message()` instead to provide a message for anyone nearby to see the hair growth emote as well.

Vimas21 replied:
If I do visible_message, how can I change the text so that it says "(USERS) hair grows to extraordinary lengths!:"?

SierraKomodo replied:
In visible message, the second parameter is what others see, and the third is what you see. The can display the name of a object/mob with `[varname.name]` or just `[varname]` - When referencing things like that be sure to use `\a` or `\the` as applicable so that improper nouns are grammatically correct (I.e. it displays "The monkey blah blah" and "Ariana Eshlian blah blah" correctly)

`visible_message(M, SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"), SPAN_NOTICE("Your hair grows to extraordinary lengths!"))`

SierraKomodo replied:
`M.visible_message(SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"))`

SierraKomodo commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,100 @@
+		M.change_facial_hair_color(red, green, blue)
+		to_chat(M, SPAN_NOTICE("Your hair suddenly changes color!"))
+
+/datum/reagent/colored_hair_dye/affect_touch(var/mob/living/carbon/human/M, var/alien, var/removed)
```
Duplicate type path - I'm assuming you meant this to be `colored_hair_dye/chaos/affect_touch(`

SierraKomodo replied:
Also the above is causing random hair colors instead of the defined hair color.

Vimas21 committed:
GitEdits

Requested by Sierra

Vimas21 committed:
More Colors

Added more colors, switched away from crayon dust, considering it doesn't exist.

SierraKomodo reviewed (dismissed):
Compile errors.

Also, I preferred the recipes that used crayon dust - The issue with that being unobtainable is easily fixed and I'm working on a PR to fix that now.

SierraKomodo commented:
#29372 has the crayon dust fix

PsiOmegaDelta commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+		var/datum/sprite_accessory/facial_hair/newbeard = /datum/sprite_accessory/facial_hair/vlongbeard
+		M.change_hair(initial(newhair.name))
+		M.change_facial_hair(initial(newbeard.name))
+		M.visible_message(SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"))
```
Not requesting a change, merely noting that you can write:
`M.visible_message(SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"), SPAN_NOTICE("Your hair grows to extraordinary lengths!"))`
`M` will then see the second message instead of the first message, which everyone else will see

Vimas21 committed:
Quick Fix

Vimas21 commented:
Sierra- I'm going to stick to my guns on switching to juice; having it be separated from crayon dust makes it both more accessible and allows more colors. There's no brown, black, or white crayon, while those colors do exist in drinks!

SierraKomodo commented:
Flagging this as ready for review in case another dev wants to take a look and decides the recipes are fine - I'm personally still meh about them.

Spookerton commented on code/modules/reagents/Chemistry-Recipes.dm:
```diff
@@ -2940,4 +2940,91 @@
+	name = "Chaos Hair Dye"
+	result = /datum/reagent/colored_hair_dye/chaos
+	required_reagents = list(/datum/reagent/hair_dye = 1, /datum/reagent/mutagen = 1)
+	result_amount = 1
```
```suggestion
	result_amount = 1

```

Spookerton commented on code/modules/reagents/Chemistry-Recipes.dm:
```diff
@@ -2940,4 +2940,91 @@
+/datum/chemical_reaction/hair_dye
+	name = "Hair Dye"
+	result = /datum/reagent/hair_dye
+	required_reagents = list(/datum/reagent/capilliumate = 1, /datum/reagent/ammonia = 1)
```
Using something made of radium and mutagen here is probably excessive for a color-carrier.
Perhaps /datum/reagent/enzyme instead?

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+	name = "Chaotic Hair Dye"
+	description = "This hair dye can be any color! Only one way to find out what kind!"
+	
+/datum/reagent/colored_hair_dye/affect_touch(var/mob/living/carbon/human/M, var/alien, var/removed)
+	metabolism=99
+	if(istype(M) && (M.species.appearance_flags & HAS_HAIR_COLOR))
+		var/red = hex2num(copytext(src.color, 2, 4))
+		var/green = hex2num(copytext(src.color, 4, 6))
+		var/blue = hex2num(copytext(src.color, 6, 8))
+		M.change_hair_color(red, green, blue)
+		M.change_facial_hair_color(red, green, blue)
+		M.visible_message(SPAN_NOTICE("\The [M]'s hair changes color!"))
```
```suggestion
/datum/reagent/colored_hair_dye/affect_touch(mob/living/carbon/human/M, alien, removed)
	if (M.h_style && M.species.appearance_flags & HAS_HAIR_COLOR)
		var/datum/sprite_accessory/hair/hair_style = GLOB.hair_styles_list[M.h_style]
		if (~hair_style.flags & HAIR_BALD)
			var/list/rgb = GetHexColors(color)
			M.change_hair_color(arglist(rgb))
			M.change_facial_hair_color(arglist(rgb))
			M.visible_message(SPAN_NOTICE("\The [M]'s hair changes color!"))
	remove_self(volume)
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+		M.change_facial_hair_color(red, green, blue)
+		M.visible_message(SPAN_NOTICE("\The [M]'s hair changes color!"))
+
+/datum/reagent/colored_hair_dye/chaos/affect_touch(var/mob/living/carbon/human/M, var/alien, var/removed)
+	if(istype(M) && (M.species.appearance_flags & HAS_HAIR_COLOR))
+		var/red = RAND_F(1,254)
+		var/green = RAND_F(1,254)
+		var/blue = RAND_F(1,254)
+		M.change_hair_color(red, green, blue)
+		M.change_facial_hair_color(red, green, blue)
+		to_chat(M, SPAN_NOTICE("Your hair suddenly changes color!"))
```
```suggestion
/datum/reagent/colored_hair_dye/chaos/affect_touch(mob/living/carbon/human/M, alien, removed)
	if (M.h_style && M.species.appearance_flags & HAS_HAIR_COLOR)
		var/datum/sprite_accessory/hair/hair_style = GLOB.hair_styles_list[M.h_style]
		if (~hair_style.flags & HAIR_BALD)
			var/red = RAND_F(1, 254)
			var/green = RAND_F(1, 254)
			var/blue = RAND_F(1, 254)
			M.change_hair_color(red, green, blue)
			M.change_facial_hair_color(red, green, blue)
			M.visible_message(SPAN_NOTICE("\The [M]'s hair changes color!"))
	remove_self(volume)
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+		M.visible_message(SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"))
+		metabolism=99
+
+/datum/reagent/capilliumate/affect_blood(var/mob/living/carbon/M, var/alien, var/removed)
+	if(alien == IS_DIONA)
+		return
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		if(prob(10))
+			to_chat(M, SPAN_NOTICE("Your tongue feels... fuzzy."))
+		if(H.stat != UNCONSCIOUS)
+			H.slurring = max(H.slurring, 10)
```
```suggestion
/datum/reagent/capilliumate/affect_blood(mob/living/carbon/M, alien, removed)
	if (alien == IS_DIONA)
		return
	if (prob(10))
		to_chat(M, SPAN_WARNING("Your tongue feels... fuzzy."))
	M.slurring = max(M.slurring, 10)
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+	heating_products = null
+	heating_point = null
+	
+/datum/reagent/capilliumate/affect_touch(var/mob/living/carbon/human/M, var/alien, var/removed)
+	if(!alien && istype(M))
+		var/datum/sprite_accessory/hair/newhair = /datum/sprite_accessory/hair/longest
+		var/datum/sprite_accessory/facial_hair/newbeard = /datum/sprite_accessory/facial_hair/vlongbeard
+		M.change_hair(initial(newhair.name))
+		M.change_facial_hair(initial(newbeard.name))
+		M.visible_message(SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"))
+		metabolism=99
```
```suggestion
/datum/reagent/capilliumate/affect_touch(mob/living/carbon/human/M, alien, removed)
	if (!alien)
		var/datum/sprite_accessory/hair/newhair = /datum/sprite_accessory/hair/longest
		var/datum/sprite_accessory/facial_hair/newbeard = /datum/sprite_accessory/facial_hair/vlongbeard
		M.change_hair(initial(newhair.name))
		M.change_facial_hair(initial(newbeard.name))
		M.visible_message(SPAN_NOTICE("\The [M]'s hair grows to extraordinary lengths!"))
	remove_self(volume)
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+
+/datum/reagent/capilliumate
+	name = "Capilliumate"
+	description = "Used across the sol system by balding men to retrieve their lost youth."
```
```suggestion
	description = "Used across the Sol system by balding men to retrieve their lost youth."
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+	reagent_state = LIQUID
+	color = "#33270b"
+	overdose = REAGENTS_OVERDOSE
+	heating_products = null
+	heating_point = null
```
```suggestion
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+	description = "Apply to your head to add some color to your life!"
+	reagent_state = LIQUID
+	taste_description = "bad choices"
+	metabolism=99
```
```suggestion
```

Spookerton commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,116 @@
+	reagent_state = LIQUID
+	color = "#b6f0ef"
+	overdose = REAGENTS_OVERDOSE
+	heating_products = null
+	heating_point = null
```
```suggestion
```

Spookerton commented:
I'm also not happy with the coloring reagent choices. Crayon betr.

SierraKomodo reviewed (changes requested):
Merge conflicts

Vimas21 committed:
Hair Rerereredux

I wasn't kidding, I will
Added Side-Effect and Chaos


GitEdits

Requested by Sierra
More Colors

Added more colors, switched away from crayon dust, considering it doesn't exist.
Quick Fix


Back to Crayons

By popular demand!

Vimas21 committed:
Merge remote-tracking branch 'Vimas21/hairchems_rererereredux' into hairchems_rererereredux

Vimas21 commented:
Oh I can feel it in my bones. I need to do it again, don't I? 

You're tearing me apart, github
